This sub-repository contains analyses relating to:
### “Hologenomics for conservation: a first test of utility.”

This project was funded by a [Royal Society of New Zealand Te Apārangi Rutherford Postdoctoral fellowship](https://royalsociety.org.nz/what-we-do/funds-and-opportunities/rutherford-foundation/rutherford-foundation-recipients/alana-alexander/) to @laninsky

#### 1. Sample selection
For the high-quality reference genomes, we included single representatives from the Māui and Hector's subspecies that had adequate tissue available (for 10x and Hi-C library construction), and that we undertook iwi consultation for.

For the larger low coverage genome resequencing/microbiome characterization, we wanted to prioritize a selection of dolphins that were confirmed to have died due to disease OR due to other sudden mortality events to act as controls (e.g. ship strike, predation, fisheries), and include in that selection at least some dolphins that had also been biopsied while alive, preferably more than once (to look at how the microbiome might shift after death, and also from sampling event to sampling event while alive). The final sample selection criterion was maximizing geographic coverage to evaluate differences in hologenomic patterns between subspecies, and between geographic areas within subspecies. [After potential resequencing Hector's individuals were identified](https://github.com/laninsky/project_logs/blob/master/hectors_hologenome/sample_selection.R), we carried out iwi consultation at the rohe level and only included samples where no concerns were raised about including them in our study.

#### 2. Metabarcoding
